internal/state/snapshot.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package state
+
+import (
+	"slices"
+
+	"github.com/siderolabs/gen/xslices"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	storagepb "github.com/siderolabs/discovery-service/api/storage"
+)
+
+// ExportClusterSnapshots exports all cluster snapshots and calls the provided function for each one.
+//
+// Implements storage.Snapshotter interface.
+func (state *State) ExportClusterSnapshots(f func(snapshot *storagepb.ClusterSnapshot) error) error {
+	var err error
+
+	// reuse the same snapshotin each iteration
+	clusterSnapshot := &storagepb.ClusterSnapshot{}
+
+	state.clusters.Range(func(_ string, cluster *Cluster) bool {
+		snapshotCluster(cluster, clusterSnapshot)
+
+		err = f(clusterSnapshot)
+
+		return err == nil
+	})
+
+	return err
+}
+
+// ImportClusterSnapshots imports cluster snapshots by calling the provided function until it returns false.
+//
+// Implements storage.Snapshotter interface.
+func (state *State) ImportClusterSnapshots(f func() (*storagepb.ClusterSnapshot, bool, error)) error {
+	for {
+		clusterSnapshot, ok, err := f()
+		if err != nil {
+			return err
+		}
+
+		if !ok {
+			break
+		}
+
+		cluster := clusterFromSnapshot(clusterSnapshot)
+
+		state.clusters.Store(cluster.id, cluster)
+	}
+
+	return nil
+}
+
+func snapshotCluster(cluster *Cluster, snapshot *storagepb.ClusterSnapshot) {
+	cluster.affiliatesMu.Lock()
+	defer cluster.affiliatesMu.Unlock()
+
+	snapshot.Id = cluster.id
+
+	// reuse the same slice, resize it as needed
+	snapshot.Affiliates = slices.Grow(snapshot.Affiliates, len(cluster.affiliates))
+	snapshot.Affiliates = snapshot.Affiliates[:len(cluster.affiliates)]
+
+	i := 0
+	for _, affiliate := range cluster.affiliates {
+		if snapshot.Affiliates[i] == nil {
+			snapshot.Affiliates[i] = &storagepb.AffiliateSnapshot{}
+		}
+
+		snapshot.Affiliates[i].Id = affiliate.id
+
+		if snapshot.Affiliates[i].Expiration == nil {
+			snapshot.Affiliates[i].Expiration = &timestamppb.Timestamp{}
+		}
+
+		snapshot.Affiliates[i].Expiration.Seconds = affiliate.expiration.Unix()
```

Comment:
there should be a helper in `timestamppb` package to do conversion between protobuf and time.Duration